### PR TITLE
Added recursive mount attr test

### PR DESCRIPTION
**Pull request by higuruchi**
Added test of rro mount option that implemented in #1398.
`/tmp/` is used as the bind mount path used for testing.

Now, this PR implement rro option test only.
And if this test is correct, add other option test(rrw, rnosuid...).

- [x] rro
- [ ] rrw
- [x] rnosuid
- [ ] rsuid
- [ ] rnodev
- [ ] rdev
- [x] rnoexec
- [ ] rexec
- [ ] rnodiratime
- [ ] rdiratime
- [ ] rrelatime
- [ ] rnorelatime
- [ ] rnoatime
- [ ] ratime
- [ ] rstrictatime
- [ ] rnostrictatime
- [ ] rnosymfollow
- [ ] rsymfollow

Signed-off-by: higuruchi <fumiya2324@gmail.com>

### CODE REVIEW

File: tests/rust-integration-tests/runtimetest/src/tests.rs [hunk 1]
```diff
@@ -76,3 +79,51 @@ pub fn validate_hostname(spec: &Spec) {
...
+            do_test_mounts_recursive(d.path().as_path(), test_fn)?;
+        }
+
+        if f_type.is_file() {
+            test_fn(d.path().as_path())?;
+        }
+    }
+
+    Ok(())
+}
```
Comment (YJDoc2) [inline tests/rust-integration-tests/runtimetest/src/tests.rs:99]: 1. don't use doc-comment for function, just use `//`
2. please improve naming,  eg, use `dirs` instead of `rd`, `dir` instead of `d` and so on

File: tests/rust-integration-tests/runtimetest/src/tests.rs [hunk 2]
```diff
@@ -76,3 +79,51 @@ pub fn validate_hostname(spec: &Spec) {
...
+
+pub fn validate_mounts_recursive(spec: &Spec) {
+    if let Some(mounts) = spec.mounts() {
+        for m in mounts {
+            if let Some(options) = m.options() {
+                for o in options {
+                    match o.as_str() {
```
Comment (YJDoc2) [inline tests/rust-integration-tests/runtimetest/src/tests.rs:106]: Similarly improve naming here as well

File: tests/rust-integration-tests/runtimetest/src/tests.rs [hunk 3]
```diff
@@ -76,3 +79,51 @@ pub fn validate_hostname(spec: &Spec) {
...
+                        "rro" => {
+                            if let Err(e) =
+                                do_test_mounts_recursive(m.destination(), &|test_file_path| {
+                                    if utils::test_write_access(test_file_path.to_str().unwrap())
+                                        .is_ok()
+                                    {
+                                        // Return Err if writeable
+                                        bail!("Unexpectedl writeable");
```
Comment (YJDoc2) [inline tests/rust-integration-tests/runtimetest/src/tests.rs:114]:
```suggestion
                                        bail!("path {:?} expected to be read-only, found writable",test_file_path);
```

File: tests/rust-integration-tests/runtimetest/src/tests.rs [hunk 4]
```diff
@@ -76,3 +79,51 @@ pub fn validate_hostname(spec: &Spec) {
...
+                                    }
+                                    Ok(())
+                                })
+                            {
+                                eprintln!("mounts_recursive rro error: {}", e);
```
Comment (YJDoc2) [inline tests/rust-integration-tests/runtimetest/src/tests.rs:119]:
```suggestion
                                eprintln!("error in testing rro recursive mounting : {}", e);
```

File: tests/rust-integration-tests/runtimetest/src/tests.rs [hunk 5]
```diff
@@ -76,3 +79,51 @@ pub fn validate_hostname(spec: &Spec) {
...
+                            }
+                        }
+                        "rrw" => { /* TODO... */ }
```
Comment (YJDoc2) [inline tests/rust-integration-tests/runtimetest/src/tests.rs:122]: Otherwise test looks fine, so please add other tests in this PR as well :+1:

File: tests/rust-integration-tests/runtimetest/src/main.rs [hunk 6]
```diff
@@ -1,9 +1,29 @@
...
+enum SubCommand {
+    #[clap(name = "readonly_paths")]
+    ReadonlyPaths,
+
+    #[clap(name = "set_host_name")]
+    SetHostName,
+
+    #[clap(name = "mounts_recursive")]
+    MountsRecursive,
+}
```
Comment (YJDoc2) [inline tests/rust-integration-tests/runtimetest/src/main.rs:25]: What is the reason for adding this? I'm not saying this is a bad idea, but why do we need this?
Also, rather than adding clap just to parse this, how about , for not, `match` on argv[1], and directly compare with string? That way we won't need this whole library.

File: tests/rust-integration-tests/integration_test/src/tests/hostname/mod.rs [hunk 7]
```diff
@@ -15,7 +15,7 @@ fn create_spec(hostname: &str) -> Spec {
         )
         .process(
             ProcessBuilder::default()
-                .args(vec!["runtimetest".to_string()])
+                .args(vec!["runtimetest".to_string(), "set_host_name".to_string()])
```
Comment (YJDoc2) [inline tests/rust-integration-tests/integration_test/src/tests/hostname/mod.rs:18]: Why this change (same question as adding clap in runtimetest)?

**Review by YJDoc2 — COMMENTED** [review]
Hey, overall seems fine, there are some changes suggested and a couple of questions. Please take a look, and add tests for remaining mount types as well.

File: tests/rust-integration-tests/runtimetest/src/main.rs (hunk 6 above)
Comment (higuruchi) [reply tests/rust-integration-tests/runtimetest/src/main.rs:25]: >What is the reason for adding this? I'm not saying this is a bad idea, but why do we need this?

As the number of test items increases, the number of functions executed in the container also increases. I added it because I thought it would be better to execute only a specific function in order to finish the test quickly.
If this is not the correct way to test, revert this to its original state.

File: tests/rust-integration-tests/integration_test/src/tests/hostname/mod.rs (hunk 7 above)
Comment (higuruchi) [reply tests/rust-integration-tests/integration_test/src/tests/hostname/mod.rs:18]: Same above comment.

File: tests/rust-integration-tests/runtimetest/src/tests.rs (hunk 1 above)
Comment (higuruchi) [reply tests/rust-integration-tests/runtimetest/src/tests.rs:99]: fixed

File: tests/rust-integration-tests/runtimetest/src/tests.rs (hunk 2 above)
Comment (higuruchi) [reply tests/rust-integration-tests/runtimetest/src/tests.rs:106]: fixed

File: tests/rust-integration-tests/runtimetest/src/tests.rs (hunk 3 above)
Comment (higuruchi) [reply tests/rust-integration-tests/runtimetest/src/tests.rs:114]: fixed

File: tests/rust-integration-tests/runtimetest/src/tests.rs (hunk 4 above)
Comment (higuruchi) [reply tests/rust-integration-tests/runtimetest/src/tests.rs:119]: fixed

**Comment by YJDoc2** [comment]
Hey @higuruchi sorry I haven't looked at this in a while. Is there any issue you are facing? If you are busy and cannot work on this, it is ok too. We can add the partial tests and add another task for remaining.

If there is any issue, feel free to ping or mention!

File: tests/rust-integration-tests/runtimetest/src/main.rs (hunk 6 above)
Comment (YJDoc2) [reply tests/rust-integration-tests/runtimetest/src/main.rs:25]: Hey @higuruchi , the idea to run specific function is great,  and we should definitely do it. However, the `runtimetest` binary is statically compiled and links all linking dependencies directly into the binary. Which means if we add clap, all of clap's linking dependencies will be statically compiled into the binary, which will make it larger in size.

As we just need to match the string, and not do any fancy parsing, can we instead just do
```rust
match argv[1]{
"readonly_paths" =>{/*...*/}
"set_host_name" =>{/*...*/}
"mount_recursive" =>{/*...*/}
_ => {panic!(....)}
}
```
That way we won't need clap

File: tests/rust-integration-tests/integration_test/src/tests/hostname/mod.rs (hunk 7 above)
Comment (YJDoc2) [reply tests/rust-integration-tests/integration_test/src/tests/hostname/mod.rs:18]: Makes sense, just see reply to the above.

File: tests/rust-integration-tests/runtimetest/src/main.rs (hunk 6 above)
Comment (higuruchi) [reply tests/rust-integration-tests/runtimetest/src/main.rs:25]: That's good.
thank you for your advice

**Comment by higuruchi** [comment]
@YJDoc2 
Also, I am unlikely to be able to work on it for a few weeks.
After fixing the clap problem, can I issue the rest?

File: tests/rust-integration-tests/runtimetest/src/main.rs [hunk 8]
```diff
@@ -38,12 +19,17 @@ fn get_spec() -> Spec {
...
     let spec = get_spec();
-
-    match opts.command {
-        SubCommand::ReadonlyPaths => tests::validate_readonly_paths(&spec),
-        SubCommand::SetHostName => tests::validate_hostname(&spec),
-        SubCommand::MountsRecursive => tests::validate_mounts_recursive(&spec),
+    let args: Vec<String> = env::args().collect();
+    let execute_test = match args.get(1) {
+        Some(execute_test) => execute_test.to_string(),
+        None => return eprintln!("error due to execute test flag not found"),
```
Comment (YJDoc2) [inline tests/rust-integration-tests/runtimetest/src/main.rs:26]: nitpick : instead of `test flag` can we make it `test name` as it is not technically a flag . Same for the error message below this.

**Comment by YJDoc2** [comment]
Hey @higuruchi No worries. If you are fine with it , we can merge this now, and when you  get time you can come back and work on the rest if you want. In the meantime, if anyone else picks up the missing cases, we can progress that way too.
Thanks for setting up the basic test and doing all this work :heart:

**Comment by higuruchi** [comment]
@YJDoc2 
> If you are fine with it , we can merge this now
Please merge if there is no problem.
And i will add more tests as time permits.

**Comment by YJDoc2** [comment]
Hey @higuruchi , sure, the PR seems mostly fine , its just that I didn't have time to take a final look.  I'll try to merge this today.
>Please merge if there is no problem.

**Review by YJDoc2 — APPROVED** [review]
lgtm :+1:

**Comment by YJDoc2** [comment]
Hey @higuruchi , I'm merging this. Thank you for the contribution !

**Comment by higuruchi** [comment]
@YJDoc2 May I raise an issue showing the rest of the tests?
